traitlets/traitlets.py

```diff
@@ -2386,296 +2362,311 @@ def info_rst(self):
             return value
-        for v in value:
+        if len(self._traits) != len(value):
+            info = "a length %i" % len(self._traits)
+            self.error(obj, len(value), info=info)
+        new = []
+        for t, v in zip(self._traits, value):
             try:
-                v = self._trait._validate(obj, v)
+                v = t._validate(obj, v)
             except TraitError as error:
                 self.error(obj, v, error)
             else:
-                validated.append(v)
-        return self.klass(validated)
+                new.append(v)
+        return tuple(new)
+
+
+class Sequence(Mutable, Container):
+    """A base trait for mutable container types
+    """
+
+    _trait = None
+    allow_none = True
+
+    def __init__(self, trait=None, default_value=Undefined,
+                 minlen=0, maxlen=sys.maxsize, **kwargs):
+        self._minlen, self._maxlen = minlen, maxlen
+        if default_value is Undefined and not is_trait(trait):
+            if trait is not None or kwargs.get("allow_none", False):
+                kwargs['default_value'], trait = trait, None
+        else:
+            kwargs['default_value'] = default_value
+        self._trait = trait
+        super(Sequence, self).__init__(**kwargs)
+
+        if inspect.isclass(trait) and issubclass(trait, TraitType):
+            warn("Traits should be given as instances, not types "
+                 "(for example, `Int()`, not `Int`). Passing "
+                 "types is deprecated in traitlets 4.1.",
+                 DeprecationWarning, stacklevel=3)
+            trait = trait()
+        if not isinstance(trait, TraitType) and trait is not None:
+            raise TypeError("The argument 'trait' must be %s, not %s." %
+                (describe("a", TraitType), describe("the", trait)))
+        self._trait = trait
+        super(Sequence, self).__init__(**kwargs)
 
     def class_init(self, cls, name):
         if isinstance(self._trait, TraitType):
             self._trait.class_init(cls, None)
-        super(Container, self).class_init(cls, name)
+        super(Sequence, self).class_init(cls, name)
 
     def instance_init(self, obj):
         if isinstance(self._trait, TraitType):
             self._trait.instance_init(obj)
-        super(Container, self).instance_init(obj)
-
-
-class List(Container):
-    """An instance of a Python list."""
-    klass = list
-    _cast_types = (tuple,)
-
-    def __init__(self, trait=None, default_value=None, minlen=0, maxlen=sys.maxsize, **kwargs):
-        """Create a List trait type from a list, set, or tuple.
-
-        The default value is created by doing ``list(default_value)``,
-        which creates a copy of the ``default_value``.
-
-        ``trait`` can be specified, which restricts the type of elements
-        in the container to that TraitType.
-
-        If only one arg is given and it is not a Trait, it is taken as
-        ``default_value``:
-
-        ``c = List([1, 2, 3])``
```

Comment:
All 'real' (user-facing) traits need to keep these detailed docstrings with examples. If anything, every PR should aim to *increase* the amount of docstring info we have, not decrease it.